File: apps/android/native/core/src/test/kotlin/org/indiafoss/companion/core/ScheduleFreshnessTest.kt
package org.indiafoss.companion.core

import java.io.File
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** Refresh state must not let a recent check read as recent data (#191). */
class ScheduleFreshnessTest {
    private val now = instant("2026-09-11T06:00:00Z")
    private val hour = 3_600_000L
    private val day = 24 * hour

    private fun instant(iso: String) = OffsetDateTime.parse(iso).toInstant().toEpochMilli()

    /** An ISO instant in the event's own offset, the shape upstream would send. */
    private fun iso(millis: Long) =
        OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneOffset.ofHoursMinutes(5, 30)).toString()

    private fun bundle(status: String? = null, updatedAt: String? = null) = EventBundle(
        id = "indiafoss-2026",
        name = "IndiaFOSS 2026",
        timezone = "Asia/Kolkata",
        start = "2026-09-26T09:00:00+05:30",
        end = "2026-09-27T17:00:00+05:30",
        sourceMetadata = SourceMetadata(scheduleStatus = status, sourceUpdatedAt = updatedAt),
    )

    @Test
    fun `anchors an unzoned upstream timestamp to the event offset`() {
        assertEquals(
            instant("2026-09-10T06:35:56.522Z"),
            ScheduleFreshness.parseSourceTimestamp("2026-09-10 12:05:56.522624", "+05:30"),
        )
        assertEquals("+05:30", ScheduleFreshness.offsetOf("2026-09-26T09:00:00+05:30"))
        assertEquals("+00:00", ScheduleFreshness.offsetOf("2026-09-26T03:30:00Z"))
        assertNull(ScheduleFreshness.offsetOf("2026-09-26T09:00:00"))
        assertNull(ScheduleFreshness.parseSourceTimestamp("2026-09-10 12:05:56", null))
        assertNull(ScheduleFreshness.parseSourceTimestamp(null, "+05:30"))
        assertNull(ScheduleFreshness.parseSourceTimestamp("not a date", "+05:30"))
    }

    @Test
    fun `describes elapsed time coarsely`() {
        assertEquals("less than an hour ago", ScheduleFreshness.describeElapsed(5 * 60_000L))
        assertEquals("1 hour ago", ScheduleFreshness.describeElapsed(hour))
        assertEquals("5 hours ago", ScheduleFreshness.describeElapsed(5 * hour))
        assertEquals("1 day ago", ScheduleFreshness.describeElapsed(day))
        assertEquals("9 days ago", ScheduleFreshness.describeElapsed(9 * day))
    }

    @Test
    fun `calls a draft programme provisional`() {
        val state = ScheduleFreshness.describe(
            bundle("draft", "2026-09-10 12:05:56.522624"),
            lastCheckedAt = now - hour,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertTrue(state.provisional)
        assertTrue(state.statusLine!!.startsWith("Provisional"))
        assertEquals(ScheduleFreshness.Age.CURRENT, state.age)
    }

    @Test
    fun `says nothing about status when the bundle does not`() {
        val state = ScheduleFreshness.describe(
            bundle(null, "2026-09-11 11:00:00"),
            lastCheckedAt = now - hour,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertFalse(state.provisional)
        assertNull(state.statusLine)
    }

    @Test
    fun `buckets the import age from the bundle, not from the last check`() {
        fun age(daysAgo: Int) = ScheduleFreshness.describe(
            bundle("confirmed", iso(now - daysAgo * day)),
            lastCheckedAt = now - 60_000L,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        ).age
        assertEquals(ScheduleFreshness.Age.CURRENT, age(0))
        assertEquals(ScheduleFreshness.Age.CURRENT, age(1))
        assertEquals(ScheduleFreshness.Age.AGEING, age(3))
        assertEquals(ScheduleFreshness.Age.STALE, age(8))
    }

    @Test
    fun `a check a minute ago does not make month-old data look current`() {
        val state = ScheduleFreshness.describe(
            bundle("confirmed", iso(now - 30 * day)),
            lastCheckedAt = now - 60_000L,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertFalse(state.checkOverdue)
        assertEquals(ScheduleFreshness.Age.STALE, state.age)
        assertTrue(state.importedLine.contains("treat these times as out of date"))
        assertTrue(state.checkedLine.contains("That is when it looked, not how old the programme is."))
    }

    @Test
    fun `an undated import is not presented as fresh`() {
        val state = ScheduleFreshness.describe(
            bundle("confirmed", null),
            lastCheckedAt = now,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertEquals(ScheduleFreshness.Age.UNKNOWN, state.age)
        assertEquals("This schedule does not record when it was imported.", state.importedLine)
    }

    @Test
    fun `a bundle still coming from the APK assets says so`() {
        val seeded = ScheduleFreshness.describe(
            bundle("confirmed", iso(now - hour)),
            lastCheckedAt = now,
            source = ScheduleFreshness.SeedSource.SEED,
            now = now,
        )
        assertTrue(seeded.sourceLine!!.contains("built into this app release"))

        val refreshed = ScheduleFreshness.describe(
            bundle("confirmed", iso(now - hour)),
            lastCheckedAt = now,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertNull(refreshed.sourceLine)
    }

    @Test
    fun `says plainly when this device has never checked`() {
        val state = ScheduleFreshness.describe(
            bundle("confirmed", iso(now - hour)),
            lastCheckedAt = null,
            source = ScheduleFreshness.SeedSource.REFRESHED,
            now = now,
        )
        assertTrue(state.checkOverdue)
        assertTrue(state.checkedLine.contains("has not checked for a newer programme yet"))
    }

    @Test
    fun `is honest with no bundle at all`() {
        val state = ScheduleFreshness.describe(
            null,
            lastCheckedAt = null,
            source = ScheduleFreshness.SeedSource.NONE,
            now = now,
        )
        assertEquals(ScheduleFreshness.Age.UNKNOWN, state.age)
        assertNull(state.statusLine)
        assertEquals("No schedule is stored on this device.", state.sourceLine)
    }

    /** The packaged 2026 bundle is the one the app actually opens on. */
    @Test
    fun `the packaged bundle's own metadata drives the wording`() {
        val packaged = EventRevisionCache.decodeBundle(
            File(repoRoot(), "events/indiafoss-2026/normalized/event-bundle.json").readText(),
            "indiafoss-2026",
        )
        val state = ScheduleFreshness.describe(
            packaged,
            lastCheckedAt = null,
            source = ScheduleFreshness.SeedSource.SEED,
            now = now,
        )
        // Whatever the current import says, the UI must never be silent about it.
        assertTrue(state.importedLine.isNotBlank())
        assertEquals(packaged.sourceMetadata.scheduleStatus == "draft", state.provisional)
    }

    private fun repoRoot(): File =
        generateSequence(File(System.getProperty("user.dir"))) { it.parentFile }
            .first { File(it, "pnpm-workspace.yaml").isFile }
}

File: apps/web/src/lib/schedule-freshness.test.ts
import { describe, expect, it } from 'vitest';
import {
  bundleOffset,
  describeElapsed,
  describeScheduleFreshness,
  parseSourceTimestamp,
} from './schedule-freshness';

const START = '2026-09-26T09:00:00+05:30';
const NOW = Date.parse('2026-09-11T06:00:00Z');
const HOUR = 60 * 60_000;
const DAY = 24 * HOUR;

type Bundle = NonNullable<Parameters<typeof describeScheduleFreshness>[0]['bundle']>;

function bundle(sourceMetadata: Bundle['sourceMetadata']): Bundle {
  return { start: START, sourceMetadata };
}

describe('bundleOffset', () => {
  it('reads the offset the bundle publishes', () => {
    expect(bundleOffset(START)).toBe('+05:30');
    expect(bundleOffset('2026-09-26T03:30:00Z')).toBe('+00:00');
  });

  it('returns null rather than guessing a zone', () => {
    expect(bundleOffset('2026-09-26T09:00:00')).toBeNull();
    expect(bundleOffset(undefined)).toBeNull();
  });
});

describe('parseSourceTimestamp', () => {
  it('anchors an unzoned upstream timestamp to the event offset', () => {
    // 12:05 IST is 06:35 UTC; a browser-local reading would be wrong by hours.
    expect(parseSourceTimestamp('2026-09-10 12:05:56.522624', '+05:30')).toBe(
      '2026-09-10T06:35:56.522Z',
    );
  });

  it('keeps a timestamp that already carries a zone', () => {
    expect(parseSourceTimestamp('2026-09-10T06:35:56Z', '+05:30')).toBe('2026-09-10T06:35:56.000Z');
  });

  it('reports nothing usable rather than inventing an instant', () => {
    expect(parseSourceTimestamp(undefined, '+05:30')).toBeNull();
    expect(parseSourceTimestamp('2026-09-10 12:05:56', null)).toBeNull();
    expect(parseSourceTimestamp('not a date', '+05:30')).toBeNull();
  });
});

describe('describeElapsed', () => {
  it('stays coarse so it never looks more precise than it is', () => {
    expect(describeElapsed(5 * 60_000)).toBe('less than an hour ago');
    expect(describeElapsed(HOUR)).toBe('1 hour ago');
    expect(describeElapsed(5 * HOUR)).toBe('5 hours ago');
    expect(describeElapsed(DAY)).toBe('1 day ago');
    expect(describeElapsed(9 * DAY)).toBe('9 days ago');
  });
});

describe('describeScheduleFreshness', () => {
  it('calls a draft programme provisional', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({
        source: 'fossunited',
        normalizerVersion: '0.2.0',
        sourceUpdatedAt: '2026-09-10 12:05:56.522624',
        scheduleStatus: 'draft',
      }),
      lastCheckedAt: NOW - HOUR,
      now: NOW,
    });
    expect(state.provisional).toBe(true);
    expect(state.statusLine).toContain('Provisional');
    expect(state.age).toBe('current');
  });

  it('says nothing about publication status when the bundle does not (2025 archive)', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({
        source: 'fossunited',
        normalizerVersion: '0.1.0',
        sourceUpdatedAt: '2026-06-02 16:08:47.991809',
      }),
      lastCheckedAt: NOW - HOUR,
      now: NOW,
    });
    expect(state.provisional).toBe(false);
    expect(state.statusLine).toBeNull();
  });

  it('marks a confirmed programme', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({
        source: 'fossunited',
        normalizerVersion: '0.2.0',
        sourceUpdatedAt: '2026-09-11 11:00:00',
        scheduleStatus: 'confirmed',
      }),
      lastCheckedAt: NOW - HOUR,
      now: NOW,
    });
    expect(state.provisional).toBe(false);
    expect(state.statusLine).toBe('The organisers have marked this programme final.');
  });

  it('buckets the import age from the bundle, not from the last check', () => {
    const at = (importedDaysAgo: number) =>
      describeScheduleFreshness({
        bundle: bundle({
          source: 'fossunited',
          normalizerVersion: '0.2.0',
          sourceUpdatedAt: new Date(NOW - importedDaysAgo * DAY).toISOString(),
          scheduleStatus: 'confirmed',
        }),
        lastCheckedAt: NOW - 60_000,
        now: NOW,
      }).age;
    expect(at(0)).toBe('current');
    expect(at(1)).toBe('current');
    expect(at(3)).toBe('ageing');
    expect(at(8)).toBe('stale');
  });

  it('does not present an undated import as fresh', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({ source: 'fossunited', normalizerVersion: '0.2.0' }),
      lastCheckedAt: NOW,
      now: NOW,
    });
    expect(state.age).toBe('unknown');
    expect(state.importedLine).toBe('This schedule does not record when it was imported.');
  });

  it('keeps the device check separate from the age of the data', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({
        source: 'fossunited',
        normalizerVersion: '0.2.0',
        sourceUpdatedAt: new Date(NOW - 30 * DAY).toISOString(),
        scheduleStatus: 'confirmed',
      }),
      lastCheckedAt: NOW - 60_000,
      now: NOW,
    });
    // A check a minute ago says nothing about a month-old import.
    expect(state.checkOverdue).toBe(false);
    expect(state.age).toBe('stale');
    expect(state.checkedLine).toContain('That is when it looked, not how old the programme is.');
  });

  it('says plainly when this device has never checked', () => {
    const state = describeScheduleFreshness({
      bundle: bundle({
        source: 'fossunited',
        normalizerVersion: '0.2.0',
        sourceUpdatedAt: new Date(NOW - HOUR).toISOString(),
        scheduleStatus: 'confirmed',
      }),
      lastCheckedAt: null,
      now: NOW,
    });
    expect(state.checkOverdue).toBe(true);
    expect(state.checkedLine).toContain('has not checked for a newer programme yet');
  });

  it('is honest before any bundle is loaded', () => {
    const state = describeScheduleFreshness({ bundle: null, lastCheckedAt: null, now: NOW });
    expect(state.age).toBe('unknown');
    expect(state.statusLine).toBeNull();
  });
});
